### ISSUE TITLE
[ENHANCEMENT] Manage upgrade-able versions in Harvester clusters

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#5102]
**Is your enhancement related to a problem? Please describe.**

It's about the Harvester version sync.

**Describe the solution you'd like**

There were situations we wanted to revoke a published version due to sane reasons before. But currently, the version sync mechanism in Harvester can't do this even if we remove a version in the upgrade responder. We need a way to handle this.

Another need is to prohibit upgrading from a non-release previous minor version. For example, say we now release v1.3.0 and set the minimum version to v1.2.2. But later we released v1.2.3 and found it's not possible to upgrade from v1.2.3 to v1.3.0, we want the v1.3.0 version to vanish in the user's cluster (The user can still see v1.3.0 in the versions list because v1.2.3 meets the minimum version of v1.3.0).

**Describe alternatives you've considered**

**Additional context**

**Comment by ibrokethecloud** [comment]
Test instructions

## Build upgrade-responder
* clone https://github.com/longhorn/upgrade-responder
* checkout branch v0.1.4
* run `make` to compile the responder
* a compiled binary will be available under ./bin directory

## Test version creation
* install harvester v1.4.0-rc3 to a cluster
* use the following response.json with the upgrade responder
```
{
    "versions": [
      {
        "Name": "v1.3.2",
        "ReleaseDate": "2024-09-07T10:00:00Z",
        "MinUpgradableVersion": "v1.3.1",
        "Tags": [
          "latest",
          "v1.3.2"
        ],
        "ExtraInfo": null
      },
      {
        "Name": "v1.4.0-rc3",
        "ReleaseDate": "2021-05-13T10:00:00Z",
        "MinUpgradableVersion": "v1.3.0",
        "Tags": [
          "v1.4.0-rc3",
          "dev"
        ],
        "ExtraInfo": null
      },
      {
        "Name": "v1.3.1",
        "ReleaseDate": "2024-06-14T10:00:00Z",
        "MinUpgradableVersion": "v1.2.2",
        "Tags": [
          "v1.3.1"
        ],
        "ExtraInfo": null
      }
    ],
    "requestIntervalInMinutes": 2
  }
```
* run the upgrade responder as follows
```
 ./bin/upgrade-responder --debug start --upgrade-response-config $path-to-response/response.json --geodb ./geodb/GeoLite2-City.mmdb --application-name harvester
```
* edit the harvester setting `upgrade-checker-url` to point to your hosted upgrade-responder endpoint
```
apiVersion: harvesterhci.io/v1beta1
default: https://harvester-upgrade-responder.rancher.io/v1/checkupgrade
kind: Setting
metadata:
  creationTimestamp: "2024-10-16T23:46:38Z"
  generation: 4
  name: upgrade-checker-url
  resourceVersion: "1003228"
  uid: 73411f21-b0e9-45b9-9b66-d7e7b0587aec
status: {}
value: http://endpoint:8314/v1/checkupgrade
```
* restart the harvester pods to force it to reconcile versions on start, else the default sync interval is 60 mins
* post restart a new version v1.4.0-rc3 should be available in harvester

## Test version cleanup
* edit the response.json and remove the `dev` tag from response.json for v1.4.0-rc3
```
{
    "versions": [
      {
        "Name": "v1.3.2",
        "ReleaseDate": "2024-09-07T10:00:00Z",
        "MinUpgradableVersion": "v1.3.1",
        "Tags": [
          "latest",
          "v1.3.2"
        ],
        "ExtraInfo": null
      },
      {
        "Name": "v1.4.0-rc3",
        "ReleaseDate": "2021-05-13T10:00:00Z",
        "MinUpgradableVersion": "v1.3.0",
        "Tags": [
          "v1.4.0-rc3"
        ],
        "ExtraInfo": null
      },
      {
        "Name": "v1.3.1",
        "ReleaseDate": "2024-06-14T10:00:00Z",
        "MinUpgradableVersion": "v1.2.2",
        "Tags": [
          "v1.3.1"
        ],
        "ExtraInfo": null
      }
    ],
    "requestIntervalInMinutes": 2
  }
```
* restart the upgrade responder
* restart the harvester pods to force resync of upgrade versions
* post restart the old version object for v1.4.0-rc3 should get cleaned up as it no longer meets the upgrade requirements as it is missing the `dev` tag in the response.json

**Comment by irishgordo** [comment]
@ibrokethecloud when trying to get this set up I'm running into a few issues.

- docker version: 27.3.1, build c312230
- go verison: v1.23.2
- make version: GNU Make 4.3
- Ubuntu 22.04 LTS Jammy Distro Base

VM Setup With:
```yaml
#cloud-config
password: password
chpasswd:
  expire: false
ssh_pwauth: true
package_update: true
manage_etc_hosts: true
packages:
  - apt-transport-https
  - ca-certificates
  - curl
  - cowsay
  - gnupg-agent
  - software-properties-common
  - qemu-guest-agent
write_files:
  - path: /etc/sysctl.d/enabled_ipv4_forwarding.conf
    content: |
      net.ipv4.conf.all.forwarding=1
groups:
  - docker
runcmd:
  - - systemctl
    - enable
    - --now
    - qemu-guest-agent.service
  - curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -
  - add-apt-repository "deb [arch=amd64]
    https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
  - apt-get update -y
  - apt-get install -y docker-ce docker-ce-cli containerd.io
    docker-buildx-plugin docker-compose-plugin
  - systemctl start docker
  - systemctl enable docker
system_info:
  default_user:
    groups: [ docker ]
network:
  version: 2
  ethernets:
    enp1s0:
      dhcp4: true
      nameservers:
        addresses: [1.1.1.1]
```

It seems that calling `make` is causing Go Lint to error out:
```go
make
Downloading dapper
.dapper.tmp version v0.6.0
./.dapper ci
[+] Building 37.9s (6/8)                                                                                                                                                                             docker:default
 => [internal] load build definition from Dockerfile.dapper1649334958                                                                                                                                          0.2s
 => => transferring dockerfile: 911B                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/golang:1.15-alpine                                                                                                                                          1.4s
 => [internal] load .dockerignore                                                                                                                                                                              0.2s
 => => transferring context: 2B                                                                                                                                                                                0.0s
 => [1/5] FROM docker.io/library/golang:1.15-alpine@sha256:b58c367d52e46cdedc25ec9cd74cadb14ad65e8db75b25e5ec117cdb227aa264                                                                                    9.1s
 => => resolve docker.io/library/golang:1.15-alpine@sha256:b58c367d52e46cdedc25ec9cd74cadb14ad65e8db75b25e5ec117cdb227aa264                                                                                    0.1s
 => => sha256:e4bc8fc554c31c0fb115880309eafbbdfcbeaa5259281e59b26346027eb06831 281.50kB / 281.50kB                                                                                                             0.3s
 => => sha256:b58c367d52e46cdedc25ec9cd74cadb14ad65e8db75b25e5ec117cdb227aa264 1.65kB / 1.65kB                                                                                                                 0.0s
 => => sha256:dc5787428863f5e2f832b52bf7405c191c575a9b38a19ae4d0cc641079242e63 1.36kB / 1.36kB                                                                                                                 0.0s
 => => sha256:1403af3b6d4ae3d0021f289a714d4df962b6be59a7fdecf4c6363c9b0f47c340 5.13kB / 5.13kB                                                                                                                 0.0s
 => => sha256:29291e31a76a7e560b9b7ad3cada56e8c18d50a96cca8a2573e4f4689d7aca77 2.81MB / 2.81MB                                                                                                                 0.5s
 => => sha256:803daa35ea4774c1839c77f23e37057a576d5cce3a041b2e2b5f700cf3f036b9 155B / 155B                                                                                                                     0.3s
 => => sha256:38284154e39640e51f13a639dd8dd9a8d68a41aa82f4891bac56a3a0e49f1290 106.92MB / 106.92MB                                                                                                             3.9s
 => => extracting sha256:29291e31a76a7e560b9b7ad3cada56e8c18d50a96cca8a2573e4f4689d7aca77                                                                                                                      0.1s
 => => sha256:a66f7597198af53100c7f92e143ffd3b724c17363091713fad62997b3bc06a5a 155B / 155B                                                                                                                     0.6s
 => => extracting sha256:e4bc8fc554c31c0fb115880309eafbbdfcbeaa5259281e59b26346027eb06831                                                                                                                      0.1s
 => => extracting sha256:803daa35ea4774c1839c77f23e37057a576d5cce3a041b2e2b5f700cf3f036b9                                                                                                                      0.0s
 => => extracting sha256:38284154e39640e51f13a639dd8dd9a8d68a41aa82f4891bac56a3a0e49f1290                                                                                                                      3.8s
 => => extracting sha256:a66f7597198af53100c7f92e143ffd3b724c17363091713fad62997b3bc06a5a                                                                                                                      0.0s
 => [2/5] RUN apk -U add bash git gcc musl-dev docker vim less file curl wget ca-certificates                                                                                                                 15.7s
 => ERROR [3/5] RUN go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golin  11.0s 
------                                                                                                                                                                                                              
 > [3/5] RUN go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golint &&     rm -rf /go/src /go/pkg:                                                                                                                                                                                                   
10.06 package golang.org/x/tools/go/gcexportdata                                                                                                                                                                    
10.06   imports golang.org/x/tools/internal/gcimporter                                                                                                                                                              
10.06 	imports golang.org/x/tools/go/gcexportdata: import cycle not allowed
------
Dockerfile.dapper1649334958:4
--------------------
   3 |     RUN apk -U add bash git gcc musl-dev docker vim less file curl wget ca-certificates
   4 | >>> RUN go get -d golang.org/x/lint/golint && \
   5 | >>>     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca && \
   6 | >>>     go install golang.org/x/lint/golint && \
   7 | >>>     rm -rf /go/src /go/pkg
   8 |     RUN go get -d github.com/alecthomas/gometalinter && \
--------------------
ERROR: failed to solve: process "/bin/sh -c go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golint &&     rm -rf /go/src /go/pkg" did not complete successfully: exit code: 1
FATA[0038] exit status 1                                
make: *** [Makefile:11: ci] Error 1
```
Specifically this chunk here:
```go
 => ERROR [3/5] RUN go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golin  11.0s 
------                                                                                                                                                                                                              
 > [3/5] RUN go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golint &&     rm -rf /go/src /go/pkg:                                                                                                                                                                                                   
10.06 package golang.org/x/tools/go/gcexportdata                                                                                                                                                                    
10.06   imports golang.org/x/tools/internal/gcimporter                                                                                                                                                              
10.06 	imports golang.org/x/tools/go/gcexportdata: import cycle not allowed
------
Dockerfile.dapper1649334958:4
--------------------
   3 |     RUN apk -U add bash git gcc musl-dev docker vim less file curl wget ca-certificates
   4 | >>> RUN go get -d golang.org/x/lint/golint && \
   5 | >>>     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca && \
   6 | >>>     go install golang.org/x/lint/golint && \
   7 | >>>     rm -rf /go/src /go/pkg
   8 |     RUN go get -d github.com/alecthomas/gometalinter && \
--------------------
ERROR: failed to solve: process "/bin/sh -c go get -d golang.org/x/lint/golint &&     git -C /go/src/golang.org/x/lint/golint checkout -b current 06c8688daad7faa9da5a0c2f163a3d14aac986ca &&     go install golang.org/x/lint/golint &&     rm -rf /go/src /go/pkg" did not complete successfully: exit code: 1
FATA[0038] exit status 1                                
make: *** [Makefile:11: ci] Error 1
```

Perhaps related to? 
- https://github.com/longhorn/upgrade-responder/commit/4563e8bb08a323bc0867ec556e3a2fd68541ffaa

Is there another branch besides `v0.4.1` I should use?

**Comment by irishgordo** [comment]
@ibrokethecloud , I can build this off of `master`

But I'm not sure this is the branch we'd need for testing :thinking: ?

```bash
Built longhornio/upgrade-responder:a6f6c77
INFO[0126] docker cp /go/src/github.com/longhorn/upgrade-responder/bin . 
Successfully copied 15.2MB to /home/ubuntu/upgrade-responder/.
INFO[0127] docker cp /go/src/github.com/longhorn/upgrade-responder/dist . 
Successfully copied 2.56kB to /home/ubuntu/upgrade-responder/.
INFO[0127] docker cp /go/src/github.com/longhorn/upgrade-responder/package . 
Successfully copied 77.5MB to /home/ubuntu/upgrade-responder/.
INFO[0128] docker cp /go/src/github.com/longhorn/upgrade-responder/coverage.out . 
Successfully copied 22.5kB to /home/ubuntu/upgrade-responder/.
ubuntu@upgraderesponder:~/upgrade-responder$ git log | head -n 10
commit a6f6c7736b7e420b07ae7d813765dac778ebc638
Author: PoAn Yang <poan.yang@suse.com>
Date:   Thu May 2 15:00:46 2024 +0800

    ci: copy geodb to package
    
    Signed-off-by: PoAn Yang <poan.yang@suse.com>

commit 726573038416d49d9c37d9a94e8694709a402781
Author: David Ko <dko@suse.com>
ubuntu@upgraderesponder:~/upgrade-responder$ git branch
* master

```

**Comment by ibrokethecloud** [comment]
looks like some package has changed upstream, the easiest option is to leverage the prebuilt container image.

Please try the following, and change `$HOME/response.json` to an appropriate path on your setup

```
 docker run -it --rm -p 8314:8314 -v $HOME/response.json:/mnt/response.json longhornio/upgrade-responder:v0.1.4 upgrade-responder --debug start --upgrade-response-config /mnt/response.json --application-name harvester
```

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.4-ce5e24e3-head` (24/11/11). Close this issue.

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Test can get the correct upgrade version from upgrade responder  $~~$</summary>

  1. When we set the `dev` tag on the `v1.4.0-rc5` version in the response.json

  2. The available upgrade version `v1.4.0-rc5` would exists on the upgrade version list 
  ![image](https://github.com/user-attachments/assets/fa863016-a736-47a9-ac30-9fbf9f0099e9)



 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test can clean up the prohibitted upgrade version from upgrade responder $~~$</summary>

  1. After we remove the `dev` tag on the `v1.4.0-rc5` version in the response.json 
  
  2. The upgrade dashboard would disappear on the host page, since no matching version exists in response.json

  ![image](https://github.com/user-attachments/assets/887d81fa-7305-4ccf-b7a9-bf51488a90d1)


 </details>

### Test Information
* Test Environment: Single node kvm machine
* Harvester version: `v1.4-ce5e24e3-head` (24/11/11)
* upgrade responder: 0.1.4

### Verify Steps
<details><summary> Test can get the correct upgrade version from upgrade responder  </summary>

  1. Prepare a v1.4.0-rc5 Harvester 
  3. Prepare another vm host for the upgrade responder
  4. Prepare the response.json for the upgrade responder
  5. Set `dev` tag on `v1.4.0-rc5`
  ```
  {
      "versions": [
        {
          "Name": "v1.3.2",
          "ReleaseDate": "2024-09-07T10:00:00Z",
          "MinUpgradableVersion": "v1.3.1",
          "Tags": [
            "latest",
            "v1.3.2"
          ],
          "ExtraInfo": null
        },
        {
          "Name": "v1.4.0-rc5",
          "ReleaseDate": "2024-11-16T10:00:00Z",
          "MinUpgradableVersion": "v1.3.0",
          "Tags": [
            "v1.4.0-rc5",
            "dev"
          ],
          "ExtraInfo": null
        },
        {
          "Name": "v1.3.1",
          "ReleaseDate": "2024-06-14T10:00:00Z",
          "MinUpgradableVersion": "v1.2.2",
          "Tags": [
            "v1.3.1"
          ],
          "ExtraInfo": null
        }
      ],
      "requestIntervalInMinutes": 2
    }
  ```
  6. Use contaner image command to create the longhorn upgrade responder
  ```
  sudo docker run -it --rm -p 8314:8314 -v ./response.json:/mnt/response.json longhornio/upgrade-responder:v0.1.4 upgrade-responder --debug start --upgrade-response-config /mnt/response.json --application-name harvester
  ```
  7. Edit the harvester setting upgrade-checker-url to point to your hosted upgrade-responder endpoint

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  apiVersion: harvesterhci.io/v1beta1
  default: https://harvester-upgrade-responder.rancher.io/v1/checkupgrade
  kind: Setting
  metadata:
    creationTimestamp: "2024-11-12T04:13:33Z"
    generation: 3
    name: upgrade-checker-url
    resourceVersion: "365196"
    uid: 7217631a-71ea-4a3c-9277-76315c631d06
  status: {}
  value: http://192.168.122.24:8314/v1/checkupgrade
  ```
  8. Restart the `system-upgrade-controller-xxx` pod
  9. Check the upgrade button exists
  10. Check the available upgrade version on the list

</details>


<details><summary> Test can clean up the prohibitted upgrade version from upgrade responder  </summary>

  1. Prepare a v1.4-head Harvester 
  3. Prepare another vm host for the upgrade responder
  4. Prepare the response.json for the upgrade responder
  5. Remove the `dev` tag on `v1.4.0-rc5`
  ```
  {
      "versions": [
        {
          "Name": "v1.3.2",
          "ReleaseDate": "2024-09-07T10:00:00Z",
          "MinUpgradableVersion": "v1.3.1",
          "Tags": [
            "latest",
            "v1.3.2"
          ],
          "ExtraInfo": null
        },
        {
          "Name": "v1.4.0-rc5",
          "ReleaseDate": "2024-11-16T10:00:00Z",
          "MinUpgradableVersion": "v1.3.0",
          "Tags": [
            "v1.4.0-rc5",
          ],
          "ExtraInfo": null
        },
        {
          "Name": "v1.3.1",
          "ReleaseDate": "2024-06-14T10:00:00Z",
          "MinUpgradableVersion": "v1.2.2",
          "Tags": [
            "v1.3.1"
          ],
          "ExtraInfo": null
        }
      ],
      "requestIntervalInMinutes": 2
    }
  ```
  6. Delete and recreate the upgrade responder container from image
  ```
  sudo docker run -it --rm -p 8314:8314 -v ./response.json:/mnt/response.json longhornio/upgrade-responder:v0.1.4 upgrade-responder --debug start --upgrade-response-config /mnt/response.json --application-name harvester
  ```
  7. Restart the `system-upgrade-controller-xxx` pod
  8. Wait for some time
  9. Check the upgrade button should disappear since no available upgrade version on the list

</details>
